### PR TITLE
A basic search endpoint

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

We'd like to have the option to search the ledger by tx out public keys, key images, and later on TxoIDs (and maybe other things).

### In this PR
* Initial implementation of a basic `search_ledger` endpoint.

### Test Plan
Unit test was added to cover the basic functionality.

### Future Work
* Expand search to cover more queries, such as TxoIDs.

### CODE REVIEW

File: full-service/src/json_rpc/v2/api/request.rs [hunk 1]
```diff
@@ -250,4 +250,7 @@ pub enum JsonCommandRequest {
         address: String,
     },
     version,
+    search_ledger {
+        query: String,
+    },
```
Comment (briancorbin) [inline full-service/src/json_rpc/v2/api/request.rs:255]: Nit - The requests should be in alphabetical order

File: full-service/src/json_rpc/v2/api/response.rs [hunk 2]
```diff
@@ -189,6 +190,9 @@ pub enum JsonCommandResponse {
         number: (String, String, String, String),
         commit: String,
     },
+    search_ledger {
+        results: Vec<LedgerSearchResult>,
+    },
```
Comment (briancorbin) [inline full-service/src/json_rpc/v2/api/response.rs:195]: Nit - The responses should be in alphabetical order

File: full-service/src/json_rpc/v2/api/wallet.rs [hunk 3]
```diff
@@ -1127,6 +1127,12 @@ where
             ),
             commit: env!("VERGEN_GIT_SHA").to_string(),
         },
+        JsonCommandRequest::search_ledger { query } => {
+            let results = service.search_ledger(&query).map_err(format_error)?;
+            JsonCommandResponse::search_ledger {
+                results: results.iter().map(Into::into).collect(),
+            }
+        }
```
Comment (briancorbin) [inline full-service/src/json_rpc/v2/api/wallet.rs:1135]: Nit - The implementations should be in alphabetical order

**Review by briancorbin — CHANGES_REQUESTED** [review]
So, this is already partially covered by the api request

```json
get_block {
        block_index: Option<String>,
        txo_public_key: Option<String>,
    },
```

which returns

```
get_block {
      block: Block,
      block_contents: BlockContents,
  }
```

I do like the idea of using a query parameter, but we could also add KeyImage to this request as an optional.

Thoughts?

**Comment by eranrund** [comment]
> So, this is already partially covered by the api request
> 
> ```json
> get_block {
>         block_index: Option<String>,
>         txo_public_key: Option<String>,
>     },
> ```
> 
> which returns
> 
> ```
> get_block {
>       block: Block,
>       block_contents: BlockContents,
>   }
> ```
> 
> I do like the idea of using a query parameter, but we could also add KeyImage to this request as an optional.
> 
> Thoughts?

I actually think this is a pretty non-standard way to implement the API - there should probably be separate endpoints for getting by block id / txo public key instead of piling more options to this endpoint.

**Review by briancorbin — APPROVED** [review]
Looks good for now, we can chat about how this should be implemented better for v3 API whenever that comes around. Thanks!

**Review by remoun — APPROVED** [review]
LGTM

**Pull request merged**